csharp/src/Ice/InputStream.cs

```diff
@@ -2138,179 +2166,147 @@ private string ReadTypeId(bool isIndex)
-                // string or as an index, for exceptions it's always encoded as a
+                // Read the type ID. For class slices, the type ID is encoded as a
+                // string or as an index or as a compact ID, for exceptions it's always encoded as a
                 // string.
-                if (_current.sliceType == SliceType.ClassSlice)
+                if (_current.InstanceType == InstanceType.Class)
                 {
                     // TYPE_ID_COMPACT must be checked first!
-                    if ((_current.sliceFlags & Protocol.FLAG_HAS_TYPE_ID_COMPACT) ==
+                    if ((_current.SliceFlags & Protocol.FLAG_HAS_TYPE_ID_COMPACT) ==
                         Protocol.FLAG_HAS_TYPE_ID_COMPACT)
                     {
-                        _current.typeId = "";
-                        _current.compactId = ReadSize();
+                        _current.SliceCompactId = ReadSize();
+                        _current.SliceTypeId = null;
                     }
-                    else if ((_current.sliceFlags &
+                    else if ((_current.SliceFlags &
                             (Protocol.FLAG_HAS_TYPE_ID_INDEX | Protocol.FLAG_HAS_TYPE_ID_STRING)) != 0)
                     {
-                        _current.typeId = ReadTypeId((_current.sliceFlags & Protocol.FLAG_HAS_TYPE_ID_INDEX) != 0);
-                        _current.compactId = -1;
+                        _current.SliceTypeId = ReadTypeId((_current.SliceFlags & Protocol.FLAG_HAS_TYPE_ID_INDEX) != 0);
+                        _current.SliceCompactId = null;
                     }
                     else
                     {
-                        // Only the most derived slice encodes the type ID for the compact format.
-                        _current.typeId = "";
-                        _current.compactId = -1;
+                        // Slice in compact format, without a type ID or compact ID.
+                        Debug.Assert((_current.SliceFlags & Protocol.FLAG_HAS_SLICE_SIZE) == 0);
+                        _current.SliceTypeId = null;
+                        _current.SliceCompactId = null;
                     }
                 }
                 else
                 {
-                    _current.typeId = ReadString();
-                    _current.compactId = -1;
+                    _current.SliceTypeId = ReadString();
+                    Debug.Assert(_current.SliceCompactId == null); // no compact ID for exceptions
                 }
 
                 // Read the slice size if necessary.
-                if ((_current.sliceFlags & Protocol.FLAG_HAS_SLICE_SIZE) != 0)
+                if ((_current.SliceFlags & Protocol.FLAG_HAS_SLICE_SIZE) != 0)
                 {
-                    _current.sliceSize = ReadInt();
-                    if (_current.sliceSize < 4)
+                    _current.SliceSize = ReadInt();
+                    if (_current.SliceSize < 4)
```

Comment:
We use `SliceSize` to access the value here. Below we use `SliceSize!.Value` and then `SliceSize.Value`. If find this a bit confusing but I also didn't spend much time to think whether or not there was a good reason to use these different accessors. Overall, in this specific case, I find that setting SliceSize to 0 if there's no slice size simpler.

Reply:
Switched back to plain ints for SliceSize as suggested.